pkg/raft/log.go

```diff
@@ -379,22 +379,44 @@ func (l *raftLog) lastIndex() uint64 {
 	return l.unstable.lastIndex()
 }
 
-// commitTo bumps the commit index to the given value if it is higher than the
-// current commit index.
-func (l *raftLog) commitTo(mark CommitMark) {
-	// TODO(pav-kv): it is only safe to update the commit index if our log is
-	// consistent with the mark.term leader. If the mark.term leader sees the
-	// mark.index entry as committed, all future leaders have it in the log. It is
-	// thus safe to bump the commit index to min(mark.index, lastIndex) if our
-	// accTerm >= mark.term. Do this once raftLog/unstable tracks the accTerm.
-
-	// never decrease commit
-	if l.committed < mark.Index {
-		if l.lastIndex() < mark.Index {
-			l.logger.Panicf("tocommit(%d) is out of range [lastIndex(%d)]. Was the raft log corrupted, truncated, or lost?", mark.Index, l.lastIndex())
-		}
-		l.committed = mark.Index
+// maybeCommit bumps the commit index if the given entry ID matches the log.
+func (l *raftLog) maybeCommit(id entryID) bool {
+	if id.index <= l.committed || !l.matchTerm(id) {
+		return false
+	}
+	l.committed = id.index
+	return true
+}
+
+// commitTo bumps the commit index in response to the knowledge that entries are
+// committed at the given mark. It can be a no-op if updating the commit index
+// is not safe, or the index is stale, or the entire log is already committed.
+//
+// Returns true iff the commit index has been bumped.
+func (l *raftLog) commitTo(mark CommitMark) bool {
+	// In the absence of evidence that entry in accTerm leader log (and so in our
+	// log too, even if it isn't there yet) matches the one in the mark.Term log
+	// (and so does the entire log prefix), it is only safe to bump the commit
+	// index if mark.Term <= accTerm (which implies this matching guarantee).
+	if mark.Term > l.accTerm() {
+		// NB: in this case, our log is lagging the leader's. If the leader is
+		// stable, we will eventually accept a MsgApp or snapshot which bumps
+		// accTerm and enables advancing the commit index again. By this, we have
+		// the guarantee that our commit index converges to the leader's.
```

Comment:
do we actually expect to hit this path? I'm not against adding a check here, but we should document whether this is basically an assertion failure, and consider logging an error.